// @ng-web-apis/permissions/esm2015/services/permissions.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: services/permissions.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { from, fromEvent, Observable } from 'rxjs';
import { map, shareReplay, startWith, switchMap } from 'rxjs/operators';
import { PERMISSIONS } from '../tokens/permissions';
import { PERMISSIONS_SUPPORT } from '../tokens/permissions-support';
import * as i0 from "@angular/core";
import * as i1 from "../tokens/permissions";
import * as i2 from "../tokens/permissions-support";
// @dynamic
export class PermissionsService {
    /**
     * @param {?} permissions
     * @param {?} permissionsSupported
     */
    constructor(permissions, permissionsSupported) {
        this.permissions = permissions;
        this.permissionsSupported = permissionsSupported;
    }
    /**
     * @param {?} nameOrDescriptor
     * @return {?}
     */
    state(nameOrDescriptor) {
        /** @type {?} */
        const descriptor = typeof nameOrDescriptor === 'string'
            ? { name: nameOrDescriptor }
            : nameOrDescriptor;
        return new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        subscriber => {
            if (!this.permissionsSupported) {
                subscriber.error('Permissions is not supported in your browser');
                return;
            }
            return from(this.permissions.query(descriptor))
                .pipe(switchMap((/**
             * @param {?} status
             * @return {?}
             */
            status => fromEvent(status, 'change').pipe(startWith(null), map((/**
             * @return {?}
             */
            () => status.state))))))
                .subscribe(subscriber);
        })).pipe(shareReplay({ bufferSize: 1, refCount: true }));
    }
}
PermissionsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
PermissionsService.ctorParameters = () => [
    { type: Permissions, decorators: [{ type: Inject, args: [PERMISSIONS,] }] },
    { type: Boolean, decorators: [{ type: Inject, args: [PERMISSIONS_SUPPORT,] }] }
];
/** @nocollapse */ PermissionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PermissionsService_Factory() { return new PermissionsService(i0.ɵɵinject(i1.PERMISSIONS), i0.ɵɵinject(i2.PERMISSIONS_SUPPORT)); }, token: PermissionsService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PermissionsService.prototype.permissions;
    /**
     * @type {?}
     * @private
     */
    PermissionsService.prototype.permissionsSupported;
}
