// @ng-web-apis/permissions/esm5/services/permissions.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: services/permissions.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { from, fromEvent, Observable } from 'rxjs';
import { map, shareReplay, startWith, switchMap } from 'rxjs/operators';
import { PERMISSIONS } from '../tokens/permissions';
import { PERMISSIONS_SUPPORT } from '../tokens/permissions-support';
import * as i0 from "@angular/core";
import * as i1 from "../tokens/permissions";
import * as i2 from "../tokens/permissions-support";
// @dynamic
var PermissionsService = /** @class */ (function () {
    function PermissionsService(permissions, permissionsSupported) {
        this.permissions = permissions;
        this.permissionsSupported = permissionsSupported;
    }
    /**
     * @param {?} nameOrDescriptor
     * @return {?}
     */
    PermissionsService.prototype.state = /**
     * @param {?} nameOrDescriptor
     * @return {?}
     */
    function (nameOrDescriptor) {
        var _this = this;
        /** @type {?} */
        var descriptor = typeof nameOrDescriptor === 'string'
            ? { name: nameOrDescriptor }
            : nameOrDescriptor;
        return new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            if (!_this.permissionsSupported) {
                subscriber.error('Permissions is not supported in your browser');
                return;
            }
            return from(_this.permissions.query(descriptor))
                .pipe(switchMap((/**
             * @param {?} status
             * @return {?}
             */
            function (status) {
                return fromEvent(status, 'change').pipe(startWith(null), map((/**
                 * @return {?}
                 */
                function () { return status.state; })));
            })))
                .subscribe(subscriber);
        })).pipe(shareReplay({ bufferSize: 1, refCount: true }));
    };
    PermissionsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    PermissionsService.ctorParameters = function () { return [
        { type: Permissions, decorators: [{ type: Inject, args: [PERMISSIONS,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [PERMISSIONS_SUPPORT,] }] }
    ]; };
    /** @nocollapse */ PermissionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PermissionsService_Factory() { return new PermissionsService(i0.ɵɵinject(i1.PERMISSIONS), i0.ɵɵinject(i2.PERMISSIONS_SUPPORT)); }, token: PermissionsService, providedIn: "root" });
    return PermissionsService;
}());
export { PermissionsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PermissionsService.prototype.permissions;
    /**
     * @type {?}
     * @private
     */
    PermissionsService.prototype.permissionsSupported;
}
